sw/device/lib/dif/dif_usbdev.c

```diff
@@ -173,12 +173,24 @@ static bool is_valid_toggle(dif_toggle_t val) {
   return val == kDifToggleEnabled || val == kDifToggleDisabled;
 }
 
+/**
+ * Checks if the given value is a valid endpoint number.
+ */
+OT_WARN_UNUSED_RESULT
+static uint8_t endpoint_number(uint8_t endpoint) { return (endpoint & 0x7F); }
+
+/**
+ * Checks if the given value is a valid endpoint number.
+ */
+OT_WARN_UNUSED_RESULT
+static bool is_in_endpoint(uint8_t endpoint) { return (endpoint & 0x80); }
+
 /**
  * Checks if the given value is a valid endpoint number.
  */
 OT_WARN_UNUSED_RESULT
 static bool is_valid_endpoint(uint8_t endpoint) {
-  return endpoint < USBDEV_NUM_ENDPOINTS;
+  return endpoint_number(endpoint) < USBDEV_NUM_ENDPOINTS;
```

Comment:
IIUC the MSB of `endpoint` now indicates the direction so `endpoint` is not just a plain number. I think it's a good idea to define a new type, e.g. `typedef uint8_t dif_usbdev_endpoint_t` with a comment. You may also want to consider adding bitfield definitions or public helper functions depending on usage.

Reply:
Yeah, sorry these software changes were a bit scattershot. I'll spend more time on the DIF.

We don't _have_ to do endpoints this way (with the MSB indicating endpoint direction), but it's actually quite common in USB software. This DIF appears to have made the mistake of not treating the IN endpoints as distinct, separate entities from the OUT endpoints, and I balked at making significant changes to the prototypes, haha. Except for the one case where SETUP packets to control endpoints cause STALL flags to be cleared for both the associated IN and OUT endpoints, the two endpoints aren't related.